File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/browser/ListingViewAdapter.java
package app.zxtune.ui.browser;

import android.net.Uri;
import android.text.TextUtils;
import android.util.SparseIntArray;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;

import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.recyclerview.selection.ItemDetailsLookup;
import androidx.recyclerview.selection.ItemKeyProvider;
import androidx.recyclerview.selection.Selection;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import app.zxtune.R;
import app.zxtune.databinding.BrowserListingEntryBinding;

class ListingViewAdapter extends ListAdapter<ListingEntry,
    ListingViewAdapter.ViewHolder> {

  private final SparseIntArray positionsCache;
  private final SearchFilter filter;
  @Nullable
  private Selection<Uri> selection;
  @Nullable
  private List<ListingEntry> lastContent;

  ListingViewAdapter() {
    super(new DiffUtil.ItemCallback<ListingEntry>() {
      @Override
      public boolean areItemsTheSame(ListingEntry oldItem, ListingEntry newItem) {
        return oldItem.getUri().equals(newItem.getUri());
      }

      @Override
      public boolean areContentsTheSame(ListingEntry oldItem, ListingEntry newItem) {
        return oldItem.equals(newItem);
      }
    });
    positionsCache = new SparseIntArray();
    filter = new SearchFilter();
    setHasStableIds(true);
  }

  @Override
  public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
    final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
    final BrowserListingEntryBinding binding = DataBindingUtil.inflate(inflater,
        R.layout.browser_listing_entry,
        parent, false);
    return new ViewHolder(binding);
  }

  @Override
  public void onBindViewHolder(ViewHolder holder, int position) {
    final ListingEntry entry = getItem(position);
    holder.bind(entry, isSelected(entry.getUri()));
  }

  final void setSelection(Selection<Uri> selection) {
    this.selection = selection;
  }

  private boolean isSelected(Uri uri) {
    return selection != null && selection.contains(uri);
  }

  @Override
  public void submitList(@Nullable List<ListingEntry> entries, @Nullable Runnable cb) {
    positionsCache.clear();
    if (entries == null || entries != lastContent) {
      lastContent = entries;
    } else {
      lastContent = new ArrayList<>(entries);
    }
    super.submitList(lastContent, cb);
  }

  final void setFilter(String pattern) {
    filter.filter(pattern);
  }

  @Override
  public void submitList(@Nullable List<ListingEntry> entries) {
    positionsCache.clear();
    super.submitList(entries);
  }

  @Override
  public long getItemId(int position) {
    return getItemInternalId(position);
  }

  final Uri getItemUri(int position) {
    return getItem(position).getUri();
  }

  private int getItemInternalId(int position) {
    return getItemUri(position).hashCode();
  }

  private int getItemPosition(@Nullable Uri uri) {
    if (uri == null) {
      return RecyclerView.NO_POSITION;
    }
    final int key = uri.hashCode();
    final int cached = positionsCache.get(key, RecyclerView.NO_POSITION);
    if (cached != RecyclerView.NO_POSITION) {
      return cached;
    }
    for (int pos = positionsCache.size(), lim = getItemCount(); pos < lim; ++pos) {
      final int id = getItemInternalId(pos);
      positionsCache.append(id, pos);
      if (key == id) {
        return pos;
      }
    }
    return RecyclerView.NO_POSITION;
  }

  static class ViewHolder extends RecyclerView.ViewHolder {

    private final BrowserListingEntryBinding binding;

    ViewHolder(BrowserListingEntryBinding binding) {
      super(binding.getRoot());
      this.binding = binding;
    }

    final void bind(ListingEntry entry, boolean isSelected) {
      binding.setEntry(entry);
      binding.executePendingBindings();
      itemView.setSelected(isSelected);
    }
  }

  static class KeyProvider extends ItemKeyProvider<Uri> {

    private final ListingViewAdapter adapter;

    KeyProvider(ListingViewAdapter adapter) {
      super(SCOPE_MAPPED);
      this.adapter = adapter;
    }

    @Nullable
    @Override
    public Uri getKey(int position) {
      return adapter.getItemUri(position);
    }

    @Override
    public int getPosition(Uri key) {
      return adapter.getItemPosition(key);
    }
  }

  static class HolderItemDetails extends ItemDetailsLookup.ItemDetails<Uri> {

    private final ViewHolder holder;

    HolderItemDetails(ViewHolder holder) {
      this.holder = holder;
    }


    @Override
    public int getPosition() {
      return holder.getAdapterPosition();
    }

    @Nullable
    @Override
    public Uri getSelectionKey() {
      return holder.binding.getEntry().getUri();
    }
  }

  static class DetailsLookup extends ItemDetailsLookup<Uri> {

    private final RecyclerView listing;

    DetailsLookup(RecyclerView view) {
      this.listing = view;
    }

    @Nullable
    @Override
    public ItemDetails<Uri> getItemDetails(MotionEvent e) {
      final View item = listing.findChildViewUnder(e.getX(), e.getY());
      if (item != null) {
        final ViewHolder holder = (ViewHolder) listing.getChildViewHolder(item);
        return new HolderItemDetails(holder);
      }
      return null;
    }
  }

  private class SearchFilter extends Filter {

    @Override
    protected FilterResults performFiltering(CharSequence constraint) {
      final FilterResults result = new FilterResults();
      if (lastContent == null) {
        return result;
      }
      if (TextUtils.isEmpty(constraint)) {
        result.values = lastContent;
      } else {
        final ArrayList<ListingEntry> filtered = new ArrayList<>();
        final String pattern = constraint.toString().toLowerCase().trim();
        for (ListingEntry entry : lastContent) {
          if (entry.getTitle().toLowerCase().contains(pattern) || entry.getDescription().toLowerCase().contains(pattern)) {
            filtered.add(entry);
          }
        }
        result.values = filtered;
      }
      return result;
    }

    @Override
    protected void publishResults(CharSequence constraint, FilterResults results) {
      if (results != null && results.values != null) {
        submitList((List<ListingEntry>) results.values);
      }
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/browser/ListingViewAdapter.kt
package app.zxtune.ui.browser

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import android.util.SparseIntArray
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.util.size
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.selection.ItemDetailsLookup
import androidx.recyclerview.selection.ItemDetailsLookup.ItemDetails
import androidx.recyclerview.selection.ItemKeyProvider
import androidx.recyclerview.selection.Selection
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import app.zxtune.R
import app.zxtune.coverart.BitmapLoader
import app.zxtune.coverart.CoverartProviderClient
import app.zxtune.databinding.BrowserListingEntryBinding
import app.zxtune.ui.utils.await
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.Job
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.async
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import kotlin.time.Duration.Companion.milliseconds

internal class ListingViewAdapter(ctx: Context) :
    ListAdapter<ListingEntry, ListingViewAdapter.ViewHolder>(DiffCallback()) {
    private val positionsCache = SparseIntArray()
    private lateinit var selection: Selection<Uri>
    private var lastContent: List<ListingEntry> = emptyList()

    @OptIn(ExperimentalCoroutinesApi::class)
    private val dispatcher = Dispatchers.IO.limitedParallelism(6)
    private val icons = BitmapLoader(
        "browser", ctx, maxUsedMemory = 1_048_576 * 5, readDispatcher = dispatcher
    )

    private lateinit var scope: CoroutineScope

    init {
        setHasStableIds(true)
    }

    override fun onAttachedToRecyclerView(rv: RecyclerView) {
        scope = MainScope()
    }

    override fun onDetachedFromRecyclerView(rv: RecyclerView) {
        scope.cancel()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding: BrowserListingEntryBinding = DataBindingUtil.inflate(
            inflater, R.layout.browser_listing_entry, parent, false
        )
        return ViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) =
        getItem(position).let { item ->
            holder.bind(item)
        }

    override fun onViewDetachedFromWindow(holder: ViewHolder) {
        holder.unbind()
    }

    fun setSelection(selection: Selection<Uri>) {
        this.selection = selection
    }

    private fun isSelected(uri: Uri) = selection.contains(uri)

    override fun submitList(entries: List<ListingEntry>?, cb: Runnable?) {
        positionsCache.clear()
        lastContent = if (entries !== lastContent) {
            entries ?: emptyList()
        } else {
            ArrayList(entries)
        }
        super.submitList(lastContent, cb)
    }

    override fun submitList(entries: List<ListingEntry>?) {
        positionsCache.clear()
        super.submitList(entries)
    }

    override fun getItemId(position: Int) = getItemInternalId(position).toLong()

    fun getItemUri(position: Int) = getItem(position).uri

    private fun getItemInternalId(position: Int) = getItemUri(position).hashCode()

    private fun getItemPosition(uri: Uri): Int {
        val key = uri.hashCode()
        val cached = positionsCache[key, RecyclerView.NO_POSITION]
        if (cached != RecyclerView.NO_POSITION) {
            return cached
        }
        var pos = positionsCache.size
        val lim = itemCount
        while (pos < lim) {
            val id = getItemInternalId(pos)
            positionsCache.append(id, pos)
            if (key == id) {
                return pos
            }
            ++pos
        }
        return RecyclerView.NO_POSITION
    }

    internal inner class ViewHolder(
        internal val binding: BrowserListingEntryBinding
    ) : RecyclerView.ViewHolder(binding.root) {
        private var loadImageJob: Job? = null

        fun bind(entry: ListingEntry) {
            unbind()
            with(binding) {
                this.entry = entry
                executePendingBindings()
            }
            itemView.isSelected = isSelected(entry.uri)
            when (entry.icon) {
                null, is ListingEntry.DrawableIcon -> setImage(null)
                is ListingEntry.LoadableIcon -> loadImage(entry.icon.uri)
            }
        }

        private fun loadImage(uri: Uri) {
            val imageUri = CoverartProviderClient.getIconUriFor(uri)
            if (!loadCached(imageUri)) {
                loadImageJob = loadRemote(imageUri)
            }
        }

        private fun loadCached(imageUri: Uri) = icons.getCached(imageUri)?.run {
            setImage(bitmap)
        } != null

        private fun loadRemote(imageUri: Uri) = scope.launch {
            val loaded = async(dispatcher) {
                icons.load(imageUri)
            }
            withTimeoutOrNull(15.milliseconds) { loaded.await() }?.let { image ->
                setImage(image.bitmap)
            } ?: run {
                setImage(null)
                loaded.await().bitmap?.let {
                    switchToImage(it)
                }
            }
        }

        fun unbind() {
            loadImageJob?.run {
                cancel()
                loadImageJob = null
            }
        }

        @SuppressLint("RtlHardcoded")
        private fun setImage(image: Bitmap?) = with(binding) {
            browserEntryIcon.alpha = 1f
            image?.let {
                browserEntryMainImage.run {
                    setImageBitmap(it)
                }
            }
            val useImage = image != null
            browserEntryMainImage.isVisible = useImage
            browserEntryMainIcon.isVisible = !useImage
            browserEntryAdditionalIcon.updateLayoutParams<FrameLayout.LayoutParams> {
                gravity = if (useImage) (Gravity.BOTTOM or Gravity.LEFT) else Gravity.CENTER
            }
        }

        private suspend fun switchToImage(image: Bitmap) =
            binding.browserEntryIcon.animate().alpha(0f).await().apply {
                setImage(image)
            }.alpha(1f).await()
    }

    internal class KeyProvider(private val adapter: ListingViewAdapter) :
        ItemKeyProvider<Uri>(SCOPE_MAPPED) {
        override fun getKey(position: Int) = adapter.getItemUri(position)

        override fun getPosition(key: Uri) = adapter.getItemPosition(key)
    }

    private class HolderItemDetails(private val holder: ViewHolder) : ItemDetails<Uri>() {
        override fun getPosition() = holder.bindingAdapterPosition

        override fun getSelectionKey() = holder.binding.entry?.uri
    }

    internal class DetailsLookup(private val listing: RecyclerView) : ItemDetailsLookup<Uri>() {
        override fun getItemDetails(e: MotionEvent): ItemDetails<Uri>? =
            listing.findChildViewUnder(e.x, e.y)?.let {
                val holder = listing.getChildViewHolder(it) as ViewHolder
                HolderItemDetails(holder)
            }
    }

    private class DiffCallback : DiffUtil.ItemCallback<ListingEntry>() {
        override fun areItemsTheSame(oldItem: ListingEntry, newItem: ListingEntry) =
            oldItem.uri == newItem.uri

        override fun areContentsTheSame(oldItem: ListingEntry, newItem: ListingEntry) =
            oldItem == newItem
    }
}
